### PR TITLE
[SPARK-36163][SQL] Propagate correct JDBC properties in JDBC connector provider and add "connectionProvider" option

### PR DESCRIPTION
**Pull request by sadikovi**
### What changes were proposed in this pull request?

This PR fixes two issues highlighted in https://issues.apache.org/jira/browse/SPARK-36163:
- JDBC connection provider propagates incorrect connection properties.
- Ambiguity when more than one JDBC connection provider is available.

I updated `BasicConnectionProvider` to use `jdbcOptions.asConnectionProperties` to remove JDBC data source specific options.

I also added `connectionProvider` data source option that specifies the name of the provider, e.g. `db2`, `presto`, to allow enforcing this specific provider in case of ambiguity.

### Why are the changes needed?
Users can leverage `spark.sql.sources.disabledJdbcConnProviderList` but it is cumbersome and requires them to disable all other providers which could be problematic when using ambiguous providers in two or more different JDBC queries.

### Does this PR introduce _any_ user-facing change?

Yes

PROBLEM DESCRIPTION:
This introduces new JDBC data source option `connectionProvider` that allows users to select a specific JDBC connection provider based on the short name. I updated the SQL guide doc and README.

Before this change, the only way to resolve ambiguity was SQL conf to blacklist all of the other JDBC connection providers. After this change users will be able to specify the exact connection provider they need per data source.

### How was this patch tested?

I updated the existing `ConnectionProviderSuite` and added a new `BasicConnectionProviderSuite`.

### CODE REVIEW

**Comment by sadikovi** [comment]
FYI @maropu
FYI @gaborgsomogyi

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala [hunk 1]
```diff
@@ -17,12 +17,21 @@
 
 package org.apache.spark.sql.execution.datasources.jdbc.connection
 
+import java.sql.{Connection, Driver}
 import javax.security.auth.login.Configuration
 
+import org.mockito.Mockito._
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:23]: shall we remove this import:

```
[error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:23:28: Unused import
[error] import org.mockito.Mockito._
[error]                            ^
[error] one error found
[error] (sql / Test / compileIncremental) Compilation failed
[error] Total time: 599 s (09:59), completed Jul 15, 2021 2:48:41 PM
```

File: docs/sql-data-sources-jdbc.md [hunk 2]
```diff
@@ -275,11 +275,22 @@ logging into the data sources.
     </td>
     <td>read/write</td>
   </tr>  
+
+  <tr>
+   <td><code>connectionProvider</code></td>
```
Comment (HyukjinKwon) [inline docs/sql-data-sources-jdbc.md:280]: I think this option conflicts with `spark.sql.sources.disabledJdbcConnProviderList` (see also https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md#how-to-disable-jdbc-connection-providers). This option is an allow list and that specific option is a disallow list.

Comment (HyukjinKwon) [reply docs/sql-data-sources-jdbc.md:280]: Oh, okay. I just noticed that you mentioned `spark.sql.sources.disabledJdbcConnProviderList` in the Pr description.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md [hunk 3]
```diff
@@ -46,6 +46,12 @@ so they can be turned off and can be replaced with custom implementation. All CP
 which must be unique. One can set the following configuration entry in `SparkConf` to turn off CPs:
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md:46]: Actually, this option has to be placed in `StaticSQLConf`, or should be effective in runtime. I will make a separate PR to move it for now.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProvider.scala [hunk 4]
```diff
@@ -55,21 +56,48 @@ private[jdbc] object ConnectionProvider extends Logging {
...
+        selectedProvider.getConnection(driver, options)
       } finally {
         logDebug("Restoring original security configuration")
         Configuration.setConfiguration(parent)
       }
     }
   }
 }
+
+private[jdbc] object ConnectionProvider extends ConnectionProviderBase
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProvider.scala:103]: You made this class for testing? Instead of creating an unnecessary class hierarchy, how about just an inner method for testing like this?
```

  // For testing
  private[jdbc] def doCreate(
    driver: Driver,
    providers: Seq[JdbcConnectionProvider],
    options: Map[String, String]): Connection = {
    
    ...
  }

  def create(driver: Driver, options: Map[String, String]): Connection = {
    doCreate(driver, providers, options)
  }
```

File: docs/sql-data-sources-jdbc.md [hunk 5]
```diff
@@ -275,11 +275,22 @@ logging into the data sources.
...
+   <td>(none)</td>
+   <td>
+     The short name of the JDBC connection provider to use to connect to this URL, e.g. <code>db2</code>, <code>mssql</code>.
+     Must be one of the providers loaded with the JDBC data source. Used to disambiguate when more than one provider can handle
+     the specified driver and options.
```
Comment (maropu) [inline docs/sql-data-sources-jdbc.md:285]: How about describing a relationship to `DISABLED_JDBC_CONN_PROVIDER_LIST` here?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md [hunk 6]
```diff
@@ -46,6 +46,12 @@ so they can be turned off and can be replaced with custom implementation. All CP
...
 `spark.sql.sources.disabledJdbcConnProviderList=name1,name2`.
 
+## How to enforce a specific JDBC connection provider?
+
+When more than one JDBC connection provider can handle a specific driver and options, it is possible to
+disambiguate and enforce a particular CP for the JDBC data source. One can set the DataFrameReader
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md:52]: Is this related to the reader only? how about the writer?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala [hunk 7]
```diff
@@ -45,6 +54,84 @@ class ConnectionProviderSuite extends ConnectionProviderSuiteBase with SharedSpa
     }
   }
 
+  test("Throw an error selecting from an empty list of providers on create") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:57]: Could you add some tests for the case where a user setting both `DISABLED_JDBC_CONN_PROVIDER_LIST` and `connectionProvider`?

File: docs/sql-data-sources-jdbc.md [hunk 8]
```diff
@@ -275,11 +275,22 @@ logging into the data sources.
     </td>
     <td>read/write</td>
   </tr>  
+
+  <tr>
+   <td><code>connectionProvider</code></td>
+   <td>(none)</td>
+   <td>
+     The short name of the JDBC connection provider to use to connect to this URL, e.g. <code>db2</code>, <code>mssql</code>.
```
Comment (maropu) [inline docs/sql-data-sources-jdbc.md:283]: What does `short` means here? `The name of a JDBC connection provider to use ...`?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md (hunk 3 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md:46]: https://github.com/apache/spark/pull/33381

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md (hunk 6 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md:52]: +1 on this question.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProviderSuite.scala [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      when(conn.getClientInfo()).thenReturn(props)
+      conn
+    })
+
+    val opts = Map(
+      JDBCOptions.JDBC_URL -> "jdbc:postgresql://localhost/postgres",
+      JDBCOptions.JDBC_TABLE_NAME -> "table",
+      JDBCOptions.JDBC_CONNECTION_PROVIDER -> "basic")
+    val conn = provider.getConnection(driver, opts)
+    assert(!conn.getClientInfo().containsKey(JDBCOptions.JDBC_URL))
```
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProviderSuite.scala:53]: Why JDBC_TABLE_NAME not checked any way?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala [hunk 10]
```diff
@@ -45,6 +54,84 @@ class ConnectionProviderSuite extends ConnectionProviderSuiteBase with SharedSpa
...
+    assert(err1.getMessage.contains("Empty list of JDBC connection providers"))
+
+    val err2 = intercept[IllegalArgumentException] {
+      providerBase.create(mock[Driver], Map.empty, Some("test"))
+    }
+    assert(err2.getMessage.contains("Empty list of JDBC connection providers"))
+  }
+
+  test("Throw an error when more than one provider is available on create") {
+    val provider1 = new JdbcConnectionProvider() {
```
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:74]: I have the feeling that `TestJdbcConnectionProvider(val name, val getConnectionImpl...)` class can be created in a common area. Such way several copy-paste can be eliminated, right?

**Comment by gaborgsomogyi** [comment]
Basically looks good, left some notes.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md (hunk 3 above)
Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md:46]: I will rebase, thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala (hunk 1 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:23]: Yes, I will redesign the tests to work around this issue.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProvider.scala (hunk 4 above)
Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProvider.scala:103]: I don't think it is a good idea to have a separate method. It would create confusion around how the providers list is generated.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md (hunk 6 above)
Comment (sadikovi) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md:52]: It is both. I meant a DataFrame option here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProviderSuite.scala (hunk 9 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/BasicConnectionProviderSuite.scala:53]: It was not reported as an issue so I did not check. I think JDBCOptions require `table` that is why it is in the map, I will add another assertion.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala (hunk 7 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:57]: Added

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala (hunk 10 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:74]: IMHO, I would have to provide closures for `canHandle` and `getConnection` to the class anyway, so I am not quite sure what the test wrapper class would accomplish.

**Comment by sadikovi** [comment]
Thanks for the reviews. I addressed most of the comments and left my replies to the rest of them.

**Comment by sadikovi** [comment]
I would appreciate another round of reviews, let me know if changes are okay or you have any concerns or objections - I will be more than happy to address.

**Comment by sadikovi** [comment]
"Throw an error if the selected provider is disabled" test fails in CI but passes locally. I will debug.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala (hunk 7 above)
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProviderSuite.scala:49]: Unfortunately, `DISABLED_JDBC_CONN_PROVIDER_LIST` is a static conf and `providers` is not lazy, i.e. I am not sure how to reevaluate providers in ConnectionProvider once the class has been loaded without reflection.

**Comment by sadikovi** [comment]
I think there `providers` should be lazy or even a method in ConnectionProvider to allow DISABLED_JDBC_CONN_PROVIDER_LIST to be set on SparkSession creation to be evaluated. cc @HyukjinKwon. It results in different outcome in the test due to different class loader ordering. For now, I am going to remove "Throw an error if the selected provider is disabled".

**Comment by HyukjinKwon** [comment]
Yeah I think it should better be lazy val

**Review by HyukjinKwon — APPROVED** [review]

**Comment by sadikovi** [comment]
It seems the tests passed! Are there any other comments or suggestions on the changes?

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by sadikovi** [comment]
Thank you very much @HyukjinKwon!
